packages/core/src/sync-realtime/service.ts

```diff
@@ -8,878 +8,633 @@ import {
+          else await wait(250 * 2 ** i);
+        }
+      }
+    },
+  });
+
+  const enqueue = () =>
+    _eth_getBlockByNumber(realtimeSyncService, {
+      blockTag: "latest",
+    }).then(queue.add);
+
+  const interval = setInterval(
+    enqueue,
+    realtimeSyncService.network.pollingInterval,
+  );
+
+  realtimeSyncService.kill = async () => {
+    realtimeSyncService.isKilled = true;
+    clearInterval(interval);
+    queue.pause();
+    queue.clear();
+
+    await queue.onIdle();
+  };
 
-    const newBlockNumber = hexToNumber(newBlock.number);
-    const newBlockTimestamp = hexToNumber(newBlock.timestamp);
+  // Note: this is done just for testing.
+  return enqueue().then(() => queue);
+};
 
-    this.emit("realtimeCheckpoint", {
-      ...maxCheckpoint,
-      blockTimestamp: newBlockTimestamp,
-      chainId: this.network.chainId,
-      blockNumber: newBlockNumber,
-    });
+/**
+ * 1) Determine if a reorg occurred.
+ * 2) Insert new event data into the store.
+ * 3) Determine if a new range of events has become finalized,
+ *    if so validate the local chain before removing the
+ *    finalized data.
+ *
+ * @param pendingLatestBlock Block to be injested. Must be exactly
+ * 1 block ahead of the local chain.
+ * @returns true if a reorg occurred
+ */
+export const handleBlock = async (
+  realtimeSyncService: Service,
+  { pendingLatestBlock }: { pendingLatestBlock: SyncBlock },
+) => {
+  const pendingLatestBlockNumber = hexToNumber(pendingLatestBlock.number);
+  const pendingLatestBlockTimestamp = hexToNumber(pendingLatestBlock.timestamp);
+  const latestLocalBlock = getLatestLocalBlock(realtimeSyncService);
+  realtimeSyncService.common.logger.debug({
+    service: "realtime",
+    msg: `Syncing new block ${pendingLatestBlockNumber} (network=${realtimeSyncService.network.name})`,
+  });
+
+  // Check if a reorg occurred by validating the chain of block hashes.
+  if (pendingLatestBlock.parentHash !== latestLocalBlock.hash) {
+    await handleReorg(realtimeSyncService, pendingLatestBlock);
+    return true;
+  }
 
-    this.common.metrics.ponder_realtime_latest_block_number.set(
-      { network: this.network.name },
-      newBlockNumber,
-    );
-    this.common.metrics.ponder_realtime_latest_block_timestamp.set(
-      { network: this.network.name },
-      newBlockTimestamp,
-    );
+  let pendingLogs: SyncLog[] = [];
 
-    this.common.logger.debug({
-      service: "realtime",
-      msg: `Finished syncing new head block ${hexToNumber(
-        newBlock.number,
-      )} (network=${this.network.name})`,
+  // "eth_getLogs" calls can be skipped if a negative match is given from "logsBloom".
+  const mustRequestGetLogs =
+    realtimeSyncService.hasFactorySource ||
+    pendingLatestBlock.logsBloom === zeroLogsBloom ||
+    isMatchedLogInBloomFilter({
+      bloom: pendingLatestBlock.logsBloom,
+      logFilters: realtimeSyncService.sources.map((s) => s.criteria),
```

Comment:
Does this work if you have both log filters and factories? The type is `criteria: LogFilterCriteria | FactoryCriteria` here.